include/gmsh.tcc

```diff
@@ -194,30 +202,127 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
       << "The coordinates for vertices of each element have been stored\n";
 }
 
-//! Compute material points based on the centroid
+//! Generate material points for 2D quadrilateral
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] ngauss_points Number of gauss points per coordinate
-template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::compute_material_points(unsigned ngauss_points) {
+//! GMSH element type
+template <>
+inline void GMSH<2, 4>::generate_material_points(unsigned ngauss_points) {
+
+  const unsigned Tdim = 2;
+  const unsigned Tvertices = 4;
+
+  //! Get constants from namespace
+  auto gauss_constants = element::gauss_points.find(ngauss_points)->second;
+
+  //! Create a matrix of xi from gauss points
+  //! Matrix is size npoints x Tdim
+  //! For 2D only
+  unsigned npoints = std::pow(ngauss_points, Tdim);
+  Eigen::MatrixXd xi_gauss_points(npoints, Tdim);
+  unsigned counter = 0;
+  for (unsigned i = 0; i < ngauss_points; ++i) {
+    for (unsigned j = 0; j < ngauss_points; ++j) {
+      xi_gauss_points(counter, 0) = gauss_constants.at(i);
+      xi_gauss_points(counter, 1) = gauss_constants.at(j);
+      ++counter;
+    }
+  }
+
+  Eigen::VectorXd pointsarray(Tdim);
+
+  //! global_id is the index for all points
+  unsigned global_id = 0;
+  //! material_id is the index of materialpoints
+  unsigned material_id = 0;
+  //! point_id is local index of points
+  unsigned point_id = 0;
+
+  //! Update vector of material points
+  //! Fill materialpoints_ vector for the first component
+  materialpoints_.emplace_back(std::unique_ptr<MaterialPoints<Tdim>>(
```

Comment:
`make_unique` and move to `materialpoints_`

Reply:
I think it would be better to have `make_unique` in `emplace_back`?